taichi/backends/metal/kernel_utils.h

```diff
@@ -208,6 +208,24 @@ struct CompiledKernelData {
   TI_IO_DEF(kernel_name, ctx_attribs, kernel_attribs);
 };
 
+struct CompiledKernelTmplData {
+  std::string kernel_bundle_name;
+  std::unordered_map<std::string, CompiledKernelData> kernel_tmpl_map;
+
+  TI_IO_DEF(kernel_bundle_name, kernel_tmpl_map);
+};
+
+struct CompiledFieldData {
+  std::string field_name;
+  MetalDataType dtype;
+  std::string dtype_name;
+  std::pair<int64_t, int64_t> dimension;
+  bool is_vector;
```

Comment:
Default initialize these primitive fields, e.g.

```suggestion
  bool is_vector{false};
```

Also, how about we reverse its meaning and rename it to `is_scalar`, since we have both vector and matrix?